lib/tesla/adapter/gun.ex

```diff
@@ -0,0 +1,200 @@
+if Code.ensure_loaded?(:gun) do
+  defmodule Tesla.Adapter.Gun do
+    @moduledoc """
+    Adapter for [gun] https://github.com/ninenines/gun
+
+    Remember to add `{:gun, "~> 1.3"}` to dependencies
+    Also, you need to recompile tesla after adding `:gun` dependency:
+    ```
+    mix deps.clean tesla
+    mix deps.compile tesla
+    ```
+    ### Example usage
+    ```
+    # set globally in config/config.exs
+    config :tesla, :adapter, Pleroma.Tesla.Adapter.Gun
+    # set per module
+    defmodule MyClient do
+      use Tesla
+      adapter Pleroma.Tesla.Adapter.Gun
+    end
+    ```
+
+    ### Options:
+
+    * `connect_timeout` - Connection timeout.
+    * `http_opts` - Options specific to the HTTP protocol.
+    * `http2_opts` -  Options specific to the HTTP/2 protocol.
+    * `protocols` - Ordered list of preferred protocols. Defaults: [http2, http] - for :tls, [http] - for :tcp.
+    * `retry` - Number of times Gun will try to reconnect on failure before giving up. Default: 5
```

Comment:
There is `Tesla.Middleware.Retry` middleware that works with all adapters, exposing this option might be confusing (which one should be used?)

Reply:
removed these options